res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js

```diff
@@ -0,0 +1,980 @@
+
+    output.addOutput("[Channel1]", "flx", 0x02, "B");
+    output.addOutput("[Channel2]", "flx", 0x29, "B");
+
+    output.addOutput("[Channel1]", "addTrack", 0x03, "B");
+    output.addOutput("[Channel2]", "addTrack", 0x2A, "B");
+
+    output.addOutput("[Channel1]", "grid", 0x05, "B");
+    output.addOutput("[Channel2]", "grid", 0x2C, "B");
+
+    output.addOutput("[Channel1]", "MaximizeLibrary", 0x04, "B");
+    output.addOutput("[Channel2]", "MaximizeLibrary", 0x2B, "B");
+
+    output.addOutput("[ChannelX]", "quantize", 0x3C, "B");
+    output.addOutput("[Microphone]", "talkover", 0x3D, "B");
+
+    this.controller.registerOutputPacket(output);
+
+    this.linkOutput("[Channel1]", "play_indicator", this.outputHandler);
+    this.linkOutput("[Channel2]", "play_indicator", this.outputHandler);
+
+    this.linkOutput("[Channel1]", "cue_indicator", this.outputHandler);
+    this.linkOutput("[Channel2]", "cue_indicator", this.outputHandler);
+
+    this.linkOutput("[Channel1]", "sync_enabled", this.outputHandler);
+    this.linkOutput("[Channel2]", "sync_enabled", this.outputHandler);
+
+    this.linkOutput("[Channel1]", "keylock", this.outputHandler);
+    this.linkOutput("[Channel2]", "keylock", this.outputHandler);
+
+    for (var i = 1; i <= 8; ++i) {
+        this.linkOutput("[Channel1]", "hotcue_" + i + "_enabled", this.hotcueOutputHandler);
+        this.linkOutput("[Channel2]", "hotcue_" + i + "_enabled", this.hotcueOutputHandler);
+    }
+
+    this.linkOutput("[Channel1]", "pfl", this.outputHandler);
+    this.linkOutput("[Channel2]", "pfl", this.outputHandler);
+
+    this.linkOutput("[Microphone]", "talkover", this.outputHandler);
+
+    // VuMeter
+    this.vuLeftConnection = engine.makeConnection("[Channel1]", "VuMeter", this.vuMeterHandler);
+    this.vuRightConnection = engine.makeConnection("[Channel2]", "VuMeter", this.vuMeterHandler);
+    this.clipLeftConnection = engine.makeConnection("[Channel1]", "PeakIndicator", this.peakOutputHandler);
+    this.clipRightConnection = engine.makeConnection("[Channel2]", "PeakIndicator", this.peakOutputHandler);
+
+    // Sampler callbacks
+    for (i = 1; i <= 16; ++i) {
+        this.samplerCallbacks.push(engine.makeConnection("[Sampler" + i + "]", "track_loaded", this.samplesOutputHandler));
+        this.samplerCallbacks.push(engine.makeConnection("[Sampler" + i + "]", "play_indicator", this.samplesOutputHandler));
+    }
+
+    TraktorS2MK3.lightDeck();
+};
+
+/* Helper function to link output in a short form */
+TraktorS2MK3.linkOutput = function (group, name, callback) {
+    TraktorS2MK3.controller.linkOutput(group, name, group, name, callback);
+};
+
+TraktorS2MK3.vuMeterHandler = function (value, group, key) {
+    var vuKeys = Object.keys(TraktorS2MK3.vuMeterThresholds);
+    for (var i = 0; i < vuKeys.length; ++i) {
+        // Avoid spamming HID by only sending last LED update
+        var last = (i === vuKeys.length - 1);
```

Comment:
This doesn't look right. Are you substracting 1 from a boolean? Just do:

```javascript
var value = ((i === vuKeys.length) ? 0 : -1);
```

Reply:
Since the subtraction operator has a higher precedence than the equality operator this line is ok i think. The variable last will be true in the last iteration of the loop. But i will add some additional parentheses to make this more obvious.

Reply:
Yes, that'd be nice.